Quit if native bcrypto bindings are missing

pinheadmz opened the pull request:
This addresses user concern https://github.com/bcoin-org/bcoin/issues/572
and replaces https://github.com/bcoin-org/bcrypto/pull/9

Test by moving or renaming the `build` directory in `bcrypto` which will force bcrypto to use nodejs crypto or proprietary crypto scripts for everything.

This PR throws an Error during `handleOpen()` that the bindings could not be found unless user explicitly sets `NODE_BACKEND=js` or `NODE_BACKEND=node`

braydonf commented:
As using JavaScript for many core operations such as `secp256k1` verification and signing is _really_ slow. I think it's a good idea to make it clear to the user that C/C++ code is not being used, as it's not practical to run a full node without it. Requiring user action to acknowledge that before proceeding can likely save debugging effort and confusion.

However, I'm not sure that this is the correct place for the check, and seems like it should be part of the build and install. For example, [secp256k1-node](https://github.com/cryptocoinjs/secp256k1-node) gives a warning during install if the build fails, and I think there have been specific environment variables to have it error in specific cases.

Also related pull requests:
- https://github.com/bcoin-org/bcoin/pull/574
- https://github.com/bcoin-org/bcoin/pull/573

tynes commented:
> However, I'm not sure that this is the correct place for the check, and seems like it should be part of the build and install.

Are you suggesting a change to `bcrypto` or having a `postInstall` npm script?

I do like the idea of this PR because users should not end up running the JS backend if they did not intend to.

pinheadmz commented:
@tynes see https://github.com/bcoin-org/bcrypto/pull/9#issuecomment-472741151
This check won't be added to `bcrypto`. I like the idea of a post-install script, but are there scenarios in which the bindings could be installed correctly but then somehow moved or corrupted in some way that the user wouldn't know at run time?

braydonf commented:
Okay, yeah there is a case where install may not have been run. Switching nodejs versions may be one of those cases. I ran into this the other day when I saw unusually very slow block verification times, and then `npm rebuild` fixed it. Would have been good to get a message.

braydonf commented on lib/node/node.js:
```diff
@@ -187,6 +188,16 @@ class Node extends EventEmitter {
+        process.env.NODE_BACKEND !== 'js' &&
+        process.env.NODE_BACKEND !== 'node'
+    ) {
+        throw new Error(
```
nit: indentation

pinheadmz replied:
does this look right?
```
    if (sha256.native !== 2 &&
        process.env.NODE_BACKEND !== 'js' &&
        process.env.NODE_BACKEND !== 'node') {
      throw new Error(
        'No bcrypto bindings. Must explicitly set `NODE_BACKEND=<js|node>`');
    }
```

braydonf replied:
Yeah, I think so.

pinheadmz replied:
:+1:

pinheadmz commented:
Just noticed this commit in hsd:

https://github.com/handshake-org/hsd/commit/4d7fdf4cec879d466a4f82d7de7cb64f9b3a4f10

Adds the same check as this PR, but in `bin/node` and quits if the primary hash functions are not built.

tynes commented on lib/node/node.js:
```diff
@@ -187,6 +188,14 @@ class Node extends EventEmitter {
+
+    // native: 0 = js, 1 = node, 2 = bindings
+    if (sha256.native !== 2 &&
+        process.env.NODE_BACKEND !== 'js' &&
```
The boolean logic here reads a bit difficult.

If its not natively built and the node backend isn't explicitly js and the node backend isn't explicitly node then throw an error. Converting to an equivalent logical statement using [De Morgan's Law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws)  that reads better in my head: if its not natively built and not (the node backend is js or the node backend is node), then throw an error.

This will need to be updated if the wasm backend ever lands.

Writing it out helped me grok it better. My only nit is that I think longer boolean statements read better when the boolean operator is in the front, things are symmetrical that way. No biggie if you don't agree :), ie 

```javascript
if (sha256.native !== 2
    && process.env.NODE_BACKEND !== 'js'
    && process.env.NODE_BACKEND !== 'node') {
  throw new Error()
}
```

pinheadmz commented:
So, the idea of putting the backend check in `bin` makes sense to me -- those files will never be run in a browser like `lib/node/node.js` where I have the check now. I'm thinking the best approach is to port the commit from hsd to bcoin's `bin/node` and `bin/spvnode` and of course change the function to `sha256()`. Thoughts?

pinheadmz commented:
Ok I replaced the commit in this PR with a port of the hsd commit. Just changed `blake2b` to `sha256`. Note that `secp256k1` is also checked, but I'm not sure how one would be built without the other.

To test:

```
$ cd bcoin/node_modules/bcrypto
$ mv build build-1
$ bcoin

Bindings for libsecp256k1 were not built.
Please build them before continuing.
```

braydonf commented:
So this has already been helpful, and warnings will appear if switching between nodejs versions and not rebuilding.
